code/go/0chain.net/chaincore/chain/state/state_context.go

```diff
@@ -400,3 +402,87 @@ func (sc *StateContext) SetStateContext(s *state.State) error {
 func (sc *StateContext) GetLatestFinalizedBlock() *block.Block {
 	return sc.getLatestFinalizedBlock()
 }
+
+type errorIndex struct {
+	err   error
+	index int
+}
+
+type GetItemFunc func(id string, balance CommonStateContextI) (interface{}, error)
+
+type rspIndex struct {
+	index int
+	item  interface{}
+}
+
+// GetItemsByIDs read items by ids from MPT concurrently and safely with consistent values
+// Note: the GetItemFunc should not return custom error that wraps the error returned from
+// StateContextI
+func GetItemsByIDs[T any](ids []string, getItem GetItemFunc, balances CommonStateContextI) ([]T, error) {
+	var (
+		itemC     = make(chan rspIndex, len(ids))
+		stateErrC = make(chan error, len(ids))
+		errC      = make(chan errorIndex, len(ids))
+		wg        sync.WaitGroup
+	)
+
+	for i, id := range ids {
+		wg.Add(1)
+		go func(idx int, id string) {
+			defer wg.Done()
+			item, err := getItem(id, balances)
+			if err != nil {
+				if err != util.ErrValueNotPresent {
+					stateErrC <- err
+					return
+				}
+
+				errC <- errorIndex{
+					err:   err,
+					index: idx,
+				}
+				return
+			}
+
+			itemC <- rspIndex{
+				index: idx,
+				item:  item,
+			}
+		}(i, id)
+	}
+	wg.Wait()
+	close(itemC)
+	close(errC)
+
+	// check internal error first
+	select {
+	case err := <-stateErrC:
+		return nil, err
+	default:
+	}
+
+	errIdxs := make([]errorIndex, 0, len(ids))
+	for ei := range errC {
+		errIdxs = append(errIdxs, ei)
```

Comment:
If `we would only return one 'value not present' error (the first one)`, then shouldn't we just grab the error form the object with lowest index value (if len(errC) > 0 )? Do we need to create an array of object and then sort the object array?

Reply:
@iamrz1 yes, good point, we should not need to sort it anymore. The sorting was done to return all errors before.